controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -324,7 +331,66 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, ownedMachines internal.FilterableMachineCollection, requireUpgrade internal.FilterableMachineCollection) (ctrl.Result, error) {
+// updateCoreDNS updates the workload cluster's kubeadm config, coredns
+// corefile and coredns deployment to match the appropriate imageTag specified
+// in the kcp DNS.
+func (r *KubeadmControlPlaneReconciler) updateCoreDNS(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane,
+) error {
+	logger := r.Log.WithValues("namespace", kcp.Namespace, "kubeadmControlPlane", kcp.Name, "cluster", cluster.Name)
+	workloadCluster, err := r.managementCluster.GetWorkloadCluster(ctx, util.ObjectKey(cluster))
+	if err != nil {
+		logger.Error(err, "failed to get remote client for workload cluster")
+		return err
+	}
+
+	currentVersion, _, _, err := workloadCluster.GetCoreDNSInfo(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to get coredns info")
+	}
+	if kcp.Spec.KubeadmConfigSpec.ClusterConfiguration == nil {
+		logger.Error(err, "ClusterConfiguration is empty, skipping coredns update")
+		return nil
+	}
+
+	if err := validateCoreDNSUpgrade(currentVersion, kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS.ImageTag); err != nil {
+		logger.Error(err, "failed coredns version validation, skipping coredns update")
+		return nil
+	}
+
+	if err := workloadCluster.UpdateCoreDNS(ctx, &kcp.Spec.KubeadmConfigSpec.ClusterConfiguration.DNS); err != nil {
```

Comment:
Crit. It shouldn't return here, especially for the case where the cluster is first initialized (creating client for workload will fail). It has to hit the code below to create the first machine.
So it may be easier to move this after the switch part.

Reply:
I moved `updateCoreDNS` above to the very end of `reconcile` so that it does go through the initialization logic.